### PR TITLE
Feature/precompile ertc 721 token uri getter

### PR DESCRIPTION
**Pull request by asiniscalchi**

### CODE REVIEW

File: precompile/erc721/src/lib.rs [hunk 1]
```diff
@@ -42,23 +49,35 @@ impl<AssetManager> Erc721Precompile<AssetManager>
...
+	fn token_uri(
+		collection_id: CollectionId,
+		handle: &mut impl PrecompileHandle,
+	) -> EvmResult<PrecompileOutput> {
+		let mut input = handle.read_input()?;
+		input.expect_arguments(1)?;
+
+		let asset_id: U256 = input.read()?;
+
+		match AssetManager::token_uri(collection_id, asset_id) {
```
[the remark this answers is not in the document]
Comment (asiniscalchi) [reply precompile/erc721/src/lib.rs]: sgtm

File: pallets/living-assets-ownership/src/traits.rs [hunk 2]
```diff
@@ -63,4 +63,6 @@ pub trait Erc721 {
 	///
 	/// The Ethereum address (`H160`) of the asset's owner or an error.
 	fn owner_of(collection_id: CollectionId, asset_id: U256) -> Result<H160, Self::Error>;
+
+	fn token_uri(collection_id: CollectionId, asset_id: U256) -> Result<Vec<u8>, Self::Error>;
```
[the remark this answers is not in the document]
Comment (asiniscalchi) [reply pallets/living-assets-ownership/src/traits.rs:81]: sgtm

File: precompile/erc721/src/lib.rs [hunk 3]
```diff
@@ -42,22 +47,30 @@ impl<AssetManager> Erc721Precompile<AssetManager>
...
+		collection_id: CollectionId,
+		handle: &mut impl PrecompileHandle,
+	) -> EvmResult<PrecompileOutput> {
 		let mut input = handle.read_input()?;
 		input.expect_arguments(1)?;
 
 		let asset_id: U256 = input.read()?;
 
-		// collection id is encoded into the contract address
-		let collection_id = match address_to_collection_id(handle.code_address()) {
```
Comment (magecnion) [inline precompile/erc721/src/lib.rs:52]: given that we don't assume the `collection_id` is the same as the contract we are instantiating anymore, could we remove `address_to_collection_id` func at all?

**Review by magecnion — CHANGES_REQUESTED** [review]

File: pallets/living-assets-ownership/src/lib.rs [hunk 4]
```diff
@@ -131,6 +132,18 @@ pub mod pallet {
 				None => Err(Error::UnexistentCollection),
 			}
 		}
+
+		fn token_uri(collection_id: CollectionId, asset_id: U256) -> Result<Vec<u8>, Self::Error> {
+			let base_uri = match CollectionBaseURI::<T>::get(collection_id) {
```
[the remark this answers is not in the document]
Comment (asiniscalchi) [reply pallets/living-assets-ownership/src/lib.rs]: that function return an `Option` , I didn't find any function that could handle it. Could you give an example ?

**Pull request merged**